codeforamerica/vita-min
# Nj 125 income editing

## Link to pivotal/JIRA issue

https://github.com/newjersey/affordability-pm/issues/125

## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

**Reminder**: merge main into this branch and get green tests before merging to main
## What was done?
- Adapted to the new XML schema for `UI/WF/SWF` to `UIWFSWF` and `UIHCWD`
- An overwrite method is now in the W2 model to capture intended behavior that combines `UIHCWD` into `UIWFSWF` if `UIWFSWF` is empty... and downstream calculators and XML for NJ1040 and NJ2450 also needed adapting for this.
- The W2 model also has validation checks for the limits for those specific types of income. The validation has to be conditional to allow for importing any info, but then turned on by the Income Review and W2 Editing controllers to not allow users to proceed without fixing.
- I decided to reuse the existing income and W2 editing controllers and forms, and put in checks for the existence of info about a given state not taxing a type of income.
- There weren't E2E tests yet specifically for just income review so I added NJ-specific tests there for this ticket's AC
## How to test?
- Use Lucky to see W2 Box 14 editing, and O'Neal Walker for seeing messaging for other income forms.
## Screenshots (for visual changes)


![image](https://github.com/user-attachments/assets/afd07c2d-6496-446c-970b-7351db96de6d)

![image](https://github.com/user-attachments/assets/713df361-2308-4573-a57f-a397ec90362c)

after trying to save values above the limits
<img width="586" alt="image" src="https://github.com/user-attachments/assets/4a029276-02e4-424c-ada8-705af85b5c2b" />

<img width="690" alt="image" src="https://github.com/user-attachments/assets/2c448a6d-6774-45b6-80e3-930509ea8057" />


![image](https://github.com/user-attachments/assets/eb2b2cdc-db21-4e0e-a5a2-309d3177dd2b)
